### PR TITLE
fix(filter-stream): Prevent emitting too small blocks

### PR DESCRIPTION
**Pull request by jhermsmeier**
Due to an error in how blocks were chunked, it was possible to emit too small or improperly sized blocks when encountering backpressure buffering.

This fixes that behavior, and ensures that all read buffers are consumed.

### CODE REVIEW

File: lib/filter-stream.js [hunk 1]
```diff
@@ -176,7 +177,7 @@ function FilterStream( blockMap, options ) {
     var buffer = Buffer.concat( this._buffers )
     var offset = 0
 
-    while( offset < buffer.length ) {
+    while( buffer.length - offset >= this.blockSize ) {
```
Comment (jviotti) [inline lib/filter-stream.js:180]: I'm not sure if this is handled elsewhere, or if its even a possible scenario, but could `buffer` not be a multiple of `this.blockSize`? For example, if `buffer` has size 16 and `this.blockSize` is 12 then:

- `buffer.length (16) - offset (0) >= this.blockSize (12)`, so this is true
- Push block  `buffer.slice( offset (0), offset (0) + this.blockSize (12) )`,
- `offset` is finally 12
- So `buffer.length (16) - offset (12) >= this.blockSize (12)`, false
- Are we ignoring the last 4 bytes of `buffer`?

File: lib/filter-stream.js [hunk 2]
```diff
@@ -190,6 +191,18 @@ function FilterStream( blockMap, options ) {
 
   },
 
+  _flush( done ) {
+
+    var buffer = Buffer.concat( this._buffers )
+    var block = Buffer.alloc( this.blockSize )
+
+    buffer.copy( block )
```
Comment (jviotti) [inline lib/filter-stream.js:199]: Also, is there a chance for `Buffer.concat( this._buffers )` to result in a buffer greater than `this.blockSize`?

**Comment by jviotti** [comment]
I added some comments about potential edge cases (ignore me if I'm missing something), but otherwise it looks great, and the tests pass :+1:

**Review by jviotti — APPROVED** [review]

File: lib/filter-stream.js [hunk 3]
```diff
@@ -176,7 +177,7 @@ function FilterStream( blockMap, options ) {
     var buffer = Buffer.concat( this._buffers )
     var offset = 0
 
-    while( offset < buffer.length ) {
+    while( buffer.length - offset >= this.blockSize ) {
       this._pushBlock( buffer.slice( offset, offset + this.blockSize ) )
       offset = offset + this.blockSize
```
Comment (lurch) [inline lib/filter-stream.js:182]: Why not `offset += this.blockSize` ?

File: lib/filter-stream.js [hunk 4]
```diff
@@ -24,6 +24,7 @@ function FilterStream( blockMap, options ) {
   /** @type {Object} options */
   this.options = options || {}
 
+  this.options.readableObjectMode = true
```
Comment (lurch) [inline lib/filter-stream.js:27]: Bit of a minor nit, but I wonder if you should explicitly create a separate `transformOptions` object to pass into `stream.Transform.call` (and selectively set specific properties), rather than blindly passing in the `options` object passed into this function? (the latter might have side-effects, if somebody adds extra properties to the `options` object that you weren't expecting?)

File: lib/filter-stream.js (hunk 3 above)
Comment (jhermsmeier) [reply lib/filter-stream.js:182]: No particular reason, just a habit from a time when `+=` wasn't optimised well by VMs – prefer the short variant?

File: lib/filter-stream.js (hunk 1 above)
Comment (jhermsmeier) [reply lib/filter-stream.js:180]: > Are we ignoring the last 4 bytes of buffer?

Nope, that part's just hidden in this diff; see lines 185-188 – the rest is stashed for the next transform (lines 170-175)

File: lib/filter-stream.js (hunk 2 above)
Comment (jhermsmeier) [reply lib/filter-stream.js:199]: Since `_transform()` will always run before a `_flush()`, and the (fixed) loop in the transform will emit block sized chunks, the leftover buffer will (theoretically) always be less than the block size.

File: lib/filter-stream.js (hunk 4 above)
Comment (jhermsmeier) [reply lib/filter-stream.js:27]: This is only specifically overridden to prevent the readable state buffer from concatenating the output blocks (dropping the block's position & address properties), but otherwise passing the options straight into the super ctor is intended, as that allows for control of things like event emitter and other stream options that wouldn't make sense to specifically map here (i.e. writableObjectMode, maxListeners, highWaterMark, ...)

File: lib/filter-stream.js (hunk 3 above)
Comment (lurch) [reply lib/filter-stream.js:182]: Doesn't matter. I find `+=` more readable, but that's probably just a personal preference.

File: lib/filter-stream.js (hunk 4 above)
Comment (lurch) [reply lib/filter-stream.js:27]: Fair enough. I guess if it was better documented, my question would have been redundant ;-)
